src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,742 @@
+      case 'width': {
+        origin.x = originValueInBlocks * info.blockWidth!;
+        copySize.width = copySizeValueInBlocks * info.blockWidth!;
+        textureSize[0] =
+          origin.x + copySize.width + textureSizePaddingValueInBlocks * info.blockWidth!;
+        break;
+      }
+      case 'height': {
+        origin.y = originValueInBlocks * info.blockHeight!;
+        copySize.height = copySizeValueInBlocks * info.blockHeight!;
+        textureSize[1] =
+          origin.y + copySize.height + textureSizePaddingValueInBlocks * info.blockHeight!;
+        break;
+      }
+      case 'depth': {
+        origin.z = originValueInBlocks;
+        copySize.depth = copySizeValueInBlocks;
+        textureSize[2] = origin.z + copySize.depth + textureSizePaddingValueInBlocks;
+        break;
+      }
+    }
+
+    const rowsPerImage = copySize.height;
+    const bytesPerRow = align(copySize.width, 256);
+    const dataSize = t.requiredBytesInCopy({ bytesPerRow, rowsPerImage }, format, copySize);
+
+    // For testing width: we copy a (_ x 2 x 2) (in texel blocks) part of a (_ x 3 x 3) (in texel blocks)
+    // texture with origin (_, 1, 1) (in texel blocks). Similarly for other coordinates.
+    t.testRun({
+      textureDataLayout: { offset: 0, bytesPerRow, rowsPerImage },
+      copySize,
+      dataSize,
+      origin,
+      textureSize,
+      format,
+      initMethod,
+      checkMethod,
+    });
+  });
+
+// Generates textureSizes which correspond to the same physicalSizeAtMipLevel
+// including virtual sizes at mip level different from the physical ones.
+function* textureSizeExpander({
+  format,
+  mipLevel,
+  texturePhysicalSizeAtMipLevelInBlocks,
+}: {
+  format: SizedTextureFormat;
+  mipLevel: number;
+  texturePhysicalSizeAtMipLevelInBlocks: GPUExtent3DDict;
+}): Generator<{ textureSize: [number, number, number] }> {
+  const info = kSizedTextureFormatInfo[format];
+  const textureSize = [
+    (texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth!) << mipLevel,
+    (texturePhysicalSizeAtMipLevelInBlocks.height * info.blockHeight!) << mipLevel,
+    texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
+  yield {
+    textureSize: textureSize as [number, number, number],
+  };
+
+  // We choose width and height of the texture so that the values are divisible by blockWidth and
+  // blockHeight respectively and so that the virtual size at mip level corresponds to the same
+  // physical size.
+  // Virtual size at mip level with modified width has width = (physical size width) - (blockWidth / 2).
+  // Virtual size at mip level with modified height has height = (physical size height) - (blockHeight / 2).
+  assert(mipLevel > 0);
+  assert(textureSize[0] > 0 && textureSize[1] > 0);
+  const modifiedWidth = textureSize[0] - (info.blockWidth! << (mipLevel - 1));
+  const modifiedHeight = textureSize[1] - (info.blockHeight! << (mipLevel - 1));
+
+  if (info.blockWidth! > 1 && modifiedWidth !== textureSize[0]) {
```

Comment:
nit: factor out into `modifyWidth` and `modifyHeight` and then use `modifyWidth && modifyHeight` below